File: gyanam-backend/scripts/ocr-images-to-text.ts
import fs from "node:fs";
import path from "node:path";
import { createWorker, PSM } from "tesseract.js";
import { getPaperArtifacts, parsePyqCliArgs } from "./pyq-paths.js";

const IMAGE_FILE_REGEX = /^page-(\d+)\.clean\.png$/i;

function getArg(args: string[], flag: string) {
  const index = args.indexOf(flag);
  return index >= 0 ? args[index + 1] : undefined;
}

function listCleanImages(
  imageDir: string,
  options: { from?: number; to?: number; oddOnly?: boolean },
) {
  if (!fs.existsSync(imageDir)) {
    throw new Error(`OCR cleaned directory not found: ${imageDir}`);
  }
  return fs
    .readdirSync(imageDir)
    .filter((name) => IMAGE_FILE_REGEX.test(name))
    .map((name) => ({
      name,
      index: Number(name.match(IMAGE_FILE_REGEX)?.[1] ?? "0"),
      fullPath: path.join(imageDir, name),
    }))
    .filter((entry) => (options.from ? entry.index >= options.from : true))
    .filter((entry) => (options.to ? entry.index <= options.to : true))
    .filter((entry) => (options.oddOnly ? entry.index % 2 === 1 : true))
    .sort((left, right) => left.index - right.index);
}

async function ocrImages(
  imageDir: string,
  outputPath: string,
  options: { from?: number; to?: number; oddOnly?: boolean },
) {
  const images = listCleanImages(imageDir, options);
  if (images.length === 0) {
    throw new Error(`No cleaned images found in ${imageDir}`);
  }

  const worker = await createWorker("eng");
  await worker.setParameters({
    tessedit_pageseg_mode: String(PSM.SINGLE_COLUMN),
    preserve_interword_spaces: "1",
  });

  const chunks: string[] = [];
  for (const image of images) {
    const result = await worker.recognize(image.fullPath);
    const text = result.data.text.trim();
    chunks.push(`-- OCR image ${image.name} --`);
    chunks.push(text);
    chunks.push("");
  }

  await worker.terminate();
  fs.writeFileSync(outputPath, `${chunks.join("\n").trim()}\n`, "utf-8");
}

async function main() {
  const args = parsePyqCliArgs();
  if (!args.year || !args.paperType) {
    throw new Error("Usage: node scripts/ocr-images-to-text.ts --year <YYYY> --paper <GS1|CSAT>");
  }

  const cliArgs = process.argv.slice(2);
  const fromArg = getArg(cliArgs, "--from");
  const toArg = getArg(cliArgs, "--to");
  const oddOnly = cliArgs.includes("--odd");
  const from = fromArg ? Number(fromArg) : undefined;
  const to = toArg ? Number(toArg) : undefined;
  const options = {
    from: Number.isInteger(from) ? from : undefined,
    to: Number.isInteger(to) ? to : undefined,
    oddOnly,
  };

  const artifacts = getPaperArtifacts({ year: args.year, paperType: args.paperType });
  const imageDir = path.join(artifacts.rawDir, "ocr-cleaned");
  await ocrImages(imageDir, artifacts.questionTextPath, options);
  console.log(`Wrote OCR text to ${artifacts.questionTextPath}`);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: gyanam-backend/scripts/python/ocr_images_to_text.py
import re
from pathlib import Path

import pytesseract
from PIL import Image


def configure_tesseract() -> None:
    candidates = [
        Path(r"C:\Program Files\Tesseract-OCR\tesseract.exe"),
        Path(r"C:\Program Files (x86)\Tesseract-OCR\tesseract.exe"),
    ]
    for candidate in candidates:
        if candidate.exists():
            pytesseract.pytesseract.tesseract_cmd = str(candidate)
            return


def build_tesseract_config(oem: int = 1, psm: int = 6) -> str:
    return f"--oem {oem} --psm {psm}"


def should_use_image(file_path: Path) -> bool:
    return re.match(r"page-\d+\.clean\.png$", file_path.name) is not None


def extract_page_number(file_path: Path) -> int:
    match = re.search(r"page-(\d+)\.clean\.png", file_path.name)
    return int(match.group(1)) if match else 0


def ocr_images_to_text(image_dir: Path, output_path: Path, lang: str = "eng") -> None:
    configure_tesseract()
    config = build_tesseract_config()
    images = [item for item in image_dir.iterdir() if item.is_file() and should_use_image(item)]
    images.sort(key=extract_page_number)

    chunks: list[str] = []
    for image_path in images:
        try:
            image = Image.open(image_path)
        except Exception as exc:  # pylint: disable=broad-except
            chunks.append(f"-- OCR image {image_path.name} (failed to open: {exc}) --")
            continue
        text = pytesseract.image_to_string(image, lang=lang, config=config)
        chunks.append(f"-- OCR image {image_path.name} --")
        chunks.append(text.strip())
        chunks.append("")

    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text("\n".join(chunks).strip() + "\n", encoding="utf-8")


def main() -> None:
    root = Path(__file__).resolve().parents[2]
    image_dir = root / "data" / "pyq" / "papers" / "2017" / "gs1" / "raw" / "ocr-cleaned"
    output_path = root / "data" / "pyq" / "papers" / "2017" / "gs1" / "raw" / "question-paper.txt"

    if not image_dir.exists():
        raise FileNotFoundError(f"OCR cleaned directory not found: {image_dir}")

    ocr_images_to_text(image_dir=image_dir, output_path=output_path)
    print(f"Wrote OCR text to {output_path}")


if __name__ == "__main__":
    main()
